### ISSUE TITLE
Dummy wurfl.h is commercially-licensed

### ISSUE THREAD

**Issue by pombredanne** [haproxy/haproxy#1948]
### Detailed Description of the Problem

The dummy wurfl.h header is commercially licensed.
See https://github.com/haproxy/haproxy/commit/cd9b9bd3e47cb2814eaab6c07d1facf222f013a9#diff-be9afa6e330d524179c2f689a4858fd7645bb3d1f05e54871b962f829f2580e8
and this header:
```
/*
 * InFuze C API - HAPROXY Dummy library version of include
 *
 * Copyright (c) ScientiaMobile, Inc.
 * http://www.scientiamobile.com
 *
 * This software package is the property of ScientiaMobile Inc. and is licensed
 * commercially according to a contract between the Licensee and ScientiaMobile Inc. (Licensor).
 * If you represent the Licensee, please refer to the licensing agreement which has been signed
 * between the two parties. If you do not represent the Licensee, you are not authorized to use
 * this software in any way.
 *
 */
```

 

### Expected Behavior

This should be a license compatible with the L/GPL instead.

### Steps to Reproduce the Behavior

1. Read the L/GPL texts
2. Read this header https://github.com/haproxy/haproxy/commit/cd9b9bd3e47cb2814eaab6c07d1facf222f013a9#diff-be9afa6e330d524179c2f689a4858fd7645bb3d1f05e54871b962f829f2580e8
3. Come to a sudden realization that there is a likely conflict ;)

### Do you have any idea what may have caused this?

Yes, this commit authored by @paulborile and committed by @wtarreau https://github.com/haproxy/haproxy/commit/cd9b9bd3e47cb2814eaab6c07d1facf222f013a9#diff-be9afa6e330d524179c2f689a4858fd7645bb3d1f05e54871b962f829f2580e8

### Do you have an idea how to solve the issue?

I think this is problematic and will be detected by license scanners as a problem. @paulborile could your dummy header be licensed under the same term as haproxy or some license compatible with the L/GPL of this repo?

This has been reported as a problem https://github.com/nexB/scancode-toolkit/issues/3162 and I am fixing by detecting this correctly, but IMHO this should be relicensed and grandfathered for the past releases

### What is your configuration?

```haproxy
I am using the head of this fine repo.
```

### Output of `haproxy -vv`

```plain
This is a source-only issue
```

### Additional Information

Why this may be good advertizing for wurfl comercial licensing, this will be detected by license scanners and will create unnecessary churn and busywork for all downstream users and redistributors, including distros.

**Comment by git001** [comment]
Well I can't see the wurfl file in the tar.gz which is downloaded in https://github.com/nexB/scancode-toolkit/issues/3162 

```
curl -vLO http://www.haproxy.org/download/2.6/src/snapshot/haproxy-ss-20221026.tar.gz
tar tfvz haproxy-ss-20221026.tar.gz|egrep cont
-rw-rw-r-- root/root       487 2022-10-25 18:28 haproxy-ss-20221026/.github/workflows/contrib.yml
-rw-rw-r-- root/root      2499 2022-10-25 18:28 haproxy-ss-20221026/examples/content-sw-sample.cfg
drwxrwxr-x root/root         0 2022-10-25 18:28 haproxy-ss-20221026/reg-tests/contrib/
-rw-rw-r-- root/root      4081 2022-10-25 18:28 haproxy-ss-20221026/reg-tests/contrib/prometheus.vtc
```

The commit which is referred in https://github.com/nexB/scancode-toolkit/issues/3162 is from "Apr 23, 2019" in the current source isn't any contrib directory as this directory was moved out from the code with commit https://github.com/haproxy/haproxy/commit/57610c694e56a6b0d55bf42f1170bad93b7b3297

**Comment by wtarreau** [comment]
Aleks, it's not in "contrib" anymore, it's been in addons/wurfl/dummy/wurfl/wurfl.h since commit 57610c69.
And indeed, the file was added with commit cd9b9bd3e4 below:
```
commit cd9b9bd3e47cb2814eaab6c07d1facf222f013a9
Author: paulborile <paul.borile@gmail.com>
Date:   Thu Apr 18 12:31:25 2019 +0200

    MINOR: contrib: dummy wurfl library
    
    This is dummy version of the Scientiamobile WURFL C API that can be used
    to successfully build/run haproxy compiled with USE_WURFL=1.
    It is marked as version 1.11.2.100 to distinguish it from any real version
    of the lib. It has no external dependencies so it should work out of the
    box by building it like this :
    
       $ make -C contrib/wurfl
    
    In order to use it, simply reference this directory as the WURFL include
    and library paths :
    
       $ make TARGET=<target> USE_WURFL=1 WURFL_INC=$PWD/contrib/wurfl WURFL_LIB
```

We need to sort that out with Paul. I guess he didn't notice either, it's probably a default header they place in every file by default but indeed it's definitely incompatible.

**Comment by wtarreau** [comment]
BTW I've just sent a mail privately to Paul now.

**Comment by wtarreau** [comment]
Paul said he will send an update shortly.

**Comment by Darlelet** [comment]
Follow-up from haproxy mailing-list (for reference):
https://www.mail-archive.com/haproxy@formilux.org/msg42917.html

**Comment by wtarreau** [comment]
Now merged in 2.8-dev, this will be backported. Thanks to all involved!

**Comment by wtarreau** [comment]
Now backported everywhere, closing!

**Comment by pombredanne** [comment]
@wtarreau thank you ++

**Comment by pombredanne** [comment]
> /*
>  * InFuze C API - HAPROXY Dummy library version of include
>  *
>  * Copyright (c) ScientiaMobile, Inc.
>  * http://www.scientiamobile.com
>  *
>  * This software package is the property of ScientiaMobile Inc. and is distributed under
>  * a dual licensing scheme:
>  *
>  * 1) commercially according to a contract between the Licensee and ScientiaMobile Inc. (Licensor).
>  *    If you represent the Licensee, please refer to the licensing agreement which has been signed
>  *    between the two parties. If you do not represent the Licensee, you are not authorized to use
>  *    this software in any way.
>  *
>  * 2) LGPL when used in the context of the HAProxy project with the purpose of testing compatibility
>  *    of HAProxy with ScientiaMobile software.
>  *
>  */
> 


@wtarreau actually I think there is still a problem there somehow. I am not sure you can put a rider on the LGPL like stated here and restrict the LGPL to a certain purpose... this would likely contradict the LGPL.

> LGPL when used in the context of the HAProxy project with the purpose of testing compatibility of HAProxy with ScientiaMobile software.

**Comment by wtarreau** [comment]
I don't know but scientiamobile contributed this change, likely after checking with their lawyers. This would need to be discussed directly with them if needed, as I'm just serving as a blind relay here :-(

With that said, that code is really ridiculous, it's just a dummy library that really is only usable in the described context (i.e. verify that we didn't break an API part that one such optional module uses). In practice this code is not usable and not supposed to be, which is why I really don't care about the license it uses. Only developers enable it to increase their build coverage.

**Comment by pombredanne** [comment]
> With that said, that code is really ridiculous, it's just a dummy library that really is only usable in the described context (i.e. verify that we didn't break an API part that one such optional module uses). In practice this code is not usable and not supposed to be, which is why I really don't care about the license it uses. Only developers enable it to increase their build coverage.

I agree, this is trivial code, but as a practical matter if was flagged as a problematic license by my license scanner which is somewhat popular... and this creates useless noise and churn for the folks that scan your code.  

As it is now, the license notice reads almost like an ad and may be an LGPL issue :|

**Comment by pombredanne** [comment]
I think this is may be in conflict with the LGPL 2.1 section 10: https://github.com/haproxy/haproxy/blob/master/doc/lgpl.txt#L369 which states

> You may not impose any further restrictions on the recipients' exercise of the rights granted herein.

**Comment by wtarreau** [comment]
But would your scanner choke on a classical dual-license ? If not, it would be worth discussing this with them, either to adjust again the wording or to relax your scanner a bit if in the end it looks legitimate (I don't know). I think it's worth contacting Luca who signed-off the license change commit: 6d6787ba7

I personally don't see the value in this restriction, but I think that it's not much different from "proprietary if you want a support contract or LGPL if you don't want" for example, and maybe it's just a matter of wording. But again I'm really not the best one to discuss this, had it been me, it would probably have been in something very relaxed like MIT, BSD2 or LGPL...

**Comment by pombredanne** [comment]
@wtarreau this is not about my scanner FWIW, but really about the LGPL conflict. I will adapt to detect anything and whatever you decide. The possible LGPL conflict is something that will make most enterprise users and discerning FOSS users choke, because it can be perceived as a field of use restriction applied to the LGPL, a no-no AFAIK. 

And I don’t think the dual-license with "commercial" choice is a problem. The problem is that you cannot restrict the LGPL and explicitly limit the license to certain products.  If it were just something like "You can use this file under the LGPL, or under a commercial/proprietary license with ScientiaMobile", that would be fine IMHO.


There is another issue wrt. the version of LGPL: with no version stated the LGPL says any version applies, which means that's LGPL-2.0-or-later. It would be best to have an explicit version 2.1 which is the version you use instead for clarity and avoid forcing your users to carry more versions of the license and eventually to make a choice.

NB: All these recommendations are to avoid making licensing an issue when using haproxy... so there is not much more that I can to help here IMHO!

@passani FYI ^

PS: @wtarreau As an aside, since we are discussing license documentation clarity, I feel that https://github.com/haproxy/haproxy/blob/master/LICENSE license statement is not entirely clear to the casual reader and this is an understatement! If you want clarity and make licensing a non-issue, may I suggest that you consider adopting conventions similar to the ones used in the Linux kernel? I modestly helped the maintainers adopt these. See https://www.kernel.org/doc/html/latest/process/license-rules.html 

See also the FSFE REUSE initiative to the same effect https://reuse.software/ for a general purpose and compatible approach.

**Comment by passani** [comment]
Hi everyone. I see that there's quite a party going on here. Let me bring ScientiaMobile's viewpoint.
Our goal is to make sure that HAProxy can be integrated with our proprietary WURFL Module while safeguarding our intellectual property. Essentially, we don't want anyone to be able to claim that parts of our proprietary software was released as LGPL, GPL or any other open source license. If Philippe or someone else has ideas on how this can be achieved, we are very open to hearing them. Releasing our code or `*.h` files as GPL/LGPL is not an option.

While I am here, a few quick comments:

>  The possible LGPL conflict is something that will make most enterprise users and discerning
>  FOSS users choke, because it can be perceived as a field of use restriction
>  applied to the LGPL, a no-no AFAIK.

Everyone is free to refer to an existing license (LGPL in this case), and state that the terms of that license apply but with a few extra restrictions.
If a company is confused about a software license, they have several options: involve their legal counsel, take some risk (calculated risk in this case) and live with the uncertainty, OR move to a commercial license.

> There is another issue wrt. the version of LGPL: with no version stated
> the LGPL says any version applies LGPL has been created explicitly to allow integrating free software
> with proprietary software. 

There is no conflict between LGPL (no matter the version) and software with whatever other license (within reason). In fact, we already stretched quite a lot when we went with HAProxy's request to review the license for our header file a couple of months ago. Going back to a single simple statement that clarifies that the WURFL header file is ScientiaMobile proprietary software would be ok for us. Thank you
